### PR TITLE
[Fix #10479] Fix a false positive for `Lint/ShadowingOuterLocalVariable` conditional statement and block variable

### LINKED ISSUE #10479: Lint/ShadowingOuterLocalVariable: a false positive case with if/else statement and block variable
## Expected behavior

No offense with `Lint/ShadowingOuterLocalVariable` cop

## Actual behavior

```ruby
W: Lint/ShadowingOuterLocalVariable: Shadowing outer local variable - email.
      list.each.map do |email|
```

`email` variable is not declared on any previous level.

## Steps to reproduce the problem

Code snippet

```ruby
  def emails_list(list, separator: ', ', short: false)
    return if list.blank?

    if short
      email = list.first
      out   = mail_to(h(email))
      out << ', &hellip;'.html_safe
    else
      list.each.map do |email|
        escaped = h(email)
        mail_to escaped, escaped, target: '_blank'
      end.uniq.join(separator).html_safe
    end
  end
```

```
rubocop --only Lint/ShadowingOuterLocalVariable
```

## RuboCop version

```
$ [bundle exec] rubocop -V
1.26.1 (using Parser 3.1.1.0, rubocop-ast 1.16.0, running on ruby 3.1.1 arm64-darwin21)
  - rubocop-performance 1.13.3
  - rubocop-rails 2.14.2
```

### PR DESCRIPTION
**Pull request by ydah**
Fixes https://github.com/rubocop/rubocop/issues/10479

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/shadowing_outer_local_variable.rb [hunk 1]
```diff
@@ -57,10 +57,20 @@ def before_declaring_variable(variable, variable_table)
...
           add_offense(variable.declaration_node, message: message)
         end
+
+        def same_conditions_node_different_branch?(variable, outer_local_variable)
+          variable_node = variable.scope.node.parent
+          outer_local_variable_node = outer_local_variable.scope.node
+
+          variable_node.conditional? &&
+            outer_local_variable_node.conditional? &&
+            variable_node == outer_local_variable_node
```
Comment (koic) [inline lib/rubocop/cop/lint/shadowing_outer_local_variable.rb:72]: The following reduces the processing that may be unnecessary.

```suggestion
          variable_node = variable.scope.node.parent
          return false unless variable_node.conditional?

          outer_local_variable_node = outer_local_variable.scope.node

          outer_local_variable_node.conditional? && variable_node == outer_local_variable_node
```

Comment (ydah) [reply lib/rubocop/cop/lint/shadowing_outer_local_variable.rb]: I see it better that way too!
Update PR. Thank you so much.

**Comment by ydah** [comment]
I updated this PR. Thanks a lot!

**Comment by koic** [comment]
Thanks!

**Pull request merged**